python/pyiceberg/table/snapshots.py

```diff
@@ -117,12 +125,369 @@ def manifests(self, io: FileIO) -> List[ManifestFile]:
+        oldest_snapshot = snapshot
+
+    return oldest_snapshot
+
+
+def oldest_ancestor_after(table: "Table", timestamp_ms: int) -> Snapshot:
+    """
+    Looks up the snapshot after a given point in time
+
+    Args:
+        table: The table
+        timestamp_ms: The timestamp in millis since the Unix epoch
+
+    Returns:
+        The snapshot after the given point in time
+
+    Raises:
+        ValueError: When there is no snapshot older than the given time
+    """
+    if last_snapshot := table.current_snapshot():
+        for snapshot in current_ancestors(table):
+            if snapshot.timestamp_ms < timestamp_ms:
+                return last_snapshot
+            elif snapshot.timestamp_ms == timestamp_ms:
+                return snapshot
+
+            last_snapshot = snapshot
+
+        if last_snapshot is not None and last_snapshot.parent_snapshot_id is None:
+            return last_snapshot
+
+    raise ValueError(f"Cannot find snapshot older than: {timestamp_ms}")
+
+
+def snapshots_ids_between(table: "Table", from_snapshot_id: int, to_snapshot_id: int) -> Iterable[int]:
+    """
+    Returns list of snapshot ids in the range - (fromSnapshotId, toSnapshotId]
+
+    This method assumes that fromSnapshotId is an ancestor of toSnapshotId.
+
+    Args:
+        table: The table
+        from_snapshot_id: The starting snapshot ID
+        to_snapshot_id: The ending snapshot ID
+
+    Returns:
+        The list of snapshot IDs that are between the given snapshot IDs
+    """
+
+    def lookup(snapshot_id: int) -> Optional[Snapshot]:
+        return table.snapshot_by_id(snapshot_id) if snapshot_id != from_snapshot_id else None
+
+    return ancestor_ids(table.snapshot_by_id(snapshot_id=to_snapshot_id), lookup)
+
+
+def ancestor_ids(latest_snapshot: Union[int, Snapshot], lookup: Callable[[int], Optional[Snapshot]]) -> Iterable[int]:
+    """
+    Returns list of the snapshot IDs of the ancestors
+
+    Args:
+        latest_snapshot: The snapshot where to start from
+        lookup: Lookup function to get the snapshot for the snapshot ID
+
+    Returns:
+        The list of snapshot IDs that are ancestor of the given snapshot
+    """
+
+    def get_id(snapshot: Snapshot) -> int:
+        return snapshot.snapshot_id
+
+    return map(get_id, ancestors_of(latest_snapshot, lookup))
+
+
+def ancestors_of(latest_snapshot: Union[int, Snapshot], lookup: Callable[[int], Optional[Snapshot]]) -> Iterable[Snapshot]:
+    """
+    Returns list of snapshot that are ancestor of the given snapshot
+
+    Args:
+        latest_snapshot: The snapshot where to start from
+        lookup: Lookup function to get the snapshot for the snapshot ID
+
+    Returns:
+        The list of snapshots that are ancestor of the given snapshot
+    """
+    if isinstance(latest_snapshot, int):
+        start = lookup(latest_snapshot)
+        if start is None:
+            raise ValueError(f"Cannot find snapshot: {latest_snapshot}")
```

Comment:
I think `lookup` will probably be `table.snapshot_by_id` and this PR changes that method to throw `ValueError` instead of returning `None`. Doesn't that conflict? Why change the behavior of `snapshot_by_id`?

Reply:
It can also be a wrapper function around `table.snapshot_by_id`: https://github.com/apache/iceberg/pull/6933/files#diff-4b189d9831bdf4c6d55c019998dd96ea7a4cd1313d2c725f582d862a8fbe0e0dR373-R374

But I agree that it is nicer to keep everything in line.